custom-ops-matrix-multiplication/benchmarks.mojo

```diff
@@ -0,0 +1,167 @@
+        shape=(M, N),
+        strides=(N, 1),
+        alignment=sizeof[dtype](),
+        address_space=AddressSpace.GENERIC,
+        exclusive=True,
+        in_lambda=None,
+        out_lambda=None,
+    )
+
+    var a_ptr = UnsafePointer[Scalar[dtype]].alloc(a_els)
+    var b_ptr = UnsafePointer[Scalar[dtype]].alloc(b_els)
+    var c_ptr = UnsafePointer[Scalar[dtype]].alloc(c_els)
+
+    var a = InputTensor[static_spec=a_spec](a_ptr, a_shape)
+    var b = InputTensor[static_spec=b_spec](b_ptr, b_shape)
+    var c = OutputTensor[static_spec=c_spec](c_ptr, c_shape)
+
+    rand(a_ptr, a_els)
+    rand(b_ptr, b_els)
+
+    var cpu_ctx_ptr = cpu_device().unsafe_ptr()
+    var bench = Bench()
+    var flops = ThroughputMeasure(BenchMetric.flops, FLOPS)
+    var elements = ThroughputMeasure(BenchMetric.elements, M * N)
+
+    @parameter
+    @always_inline
+    fn bench_cpu(mut bencher: Bencher) raises:
+        @parameter
+        @always_inline
+        fn run_bench() raises:
+            MatrixMultiplication["naive"].execute[target="cpu"](
+                c,
+                a,
+                b,
+                cpu_ctx_ptr,
+            )
+
+        bencher.iter[run_bench]()
+
+    bench.bench_function[bench_cpu](BenchId("cpu", "naive"), flops, elements)
+
+    @parameter
+    if has_nvidia_gpu_accelerator() or has_amd_gpu_accelerator():
+        var gpu_ctx = DeviceContext()
+        var a_dev = gpu_ctx.enqueue_create_buffer[dtype](a_els)
+        var b_dev = gpu_ctx.enqueue_create_buffer[dtype](b_els)
+        var c_dev = gpu_ctx.enqueue_create_buffer[dtype](c_els)
+        var c = InputTensor[static_spec=c_spec](c_dev.unsafe_ptr(), c_shape)
+        var a = OutputTensor[static_spec=a_spec](a_dev.unsafe_ptr(), a_shape)
+        var b = OutputTensor[static_spec=b_spec](b_dev.unsafe_ptr(), b_shape)
+        gpu_ctx.copy(a_dev, a_ptr)
+        gpu_ctx.copy(b_dev, b_ptr)
+
+        @parameter
+        def bench_matmul_kernel[impl: StringLiteral]():
+            @parameter
+            @always_inline
+            fn bench_gpu(mut bench: Bencher) raises:
+                @parameter
+                @always_inline
+                fn kernel_launch(gpu_ctx: DeviceContext) raises:
+                    MatrixMultiplication[impl].execute[target="gpu"](
+                        c,
+                        a,
+                        b,
+                        gpu_ctx,
+                    )
+
+                var gpu_ctx = DeviceContext()
+                bench.iter_custom[kernel_launch](gpu_ctx)
+                _ = gpu_ctx
+
+            bench.bench_function[bench_gpu](
+                BenchId("gpu", impl), flops, elements
+            )
+
+        bench_matmul_kernel["naive"]()
+        bench_matmul_kernel["coalescing"]()
+        bench_matmul_kernel["tiled"]()
+        bench_matmul_kernel["tiled_register"]()
+        bench_matmul_kernel["block_tiled"]()
+        bench_matmul_kernel["block_tiled_vectorized"]()
+        _ = gpu_ctx
```

Comment:
Are these needed at the end? the origin was recently added to the layout tensor and not sure about the rest?

Reply:
Unfortunately, I think they're still needed. We're only in the middle of transitioning these types, and when I tried removing the manual lifetime extensions on the latest nightly this still crashes on GPU execution with CUDA illegal access errors.